contracts/src/core/EigenDADisperserRegistryV2.sol

```diff
@@ -0,0 +1,205 @@
+import {EigenDADisperserRegistryStorageV2} from "src/core/EigenDADisperserRegistryStorageV2.sol";
+
+/// @title Registry for EigenDA disperser info V2
+/// @author Layr Labs, Inc.
+contract EigenDADisperserRegistryV2 is
+    OwnableUpgradeable,
+    EIP712Upgradeable,
+    EigenDADisperserRegistryStorageV2,
+    IEigenDADisperserRegistryV2
+{
+    using SignatureCheckerUpgradeable for address;
+    using EnumerableSetUpgradeable for *;
+
+    /// -----------------------------------------------------------------------
+    /// Initialization
+    /// -----------------------------------------------------------------------
+
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(address _initialOwner) external initializer {
+        __Ownable_init();
+        __EIP712_init("EigenDADisperserRegistry", "2");
+        _transferOwnership(_initialOwner);
+    }
+
+    /// -----------------------------------------------------------------------
+    /// External Logic
+    /// -----------------------------------------------------------------------
+
+    /// @inheritdoc IEigenDADisperserRegistryV2
+    function registerDisperser(address disperser, string memory relayURL)
+        external
+        virtual
+        returns (uint32 disperserId)
+    {
+        // Increment and assign the next available disperserId, starting at 1 (not 0).
+        disperserId = ++totalRegistrations; // Monotonic increasing.
+
+        EigenDATypesV2.DisperserInfoV2 storage disperserInfo = _disperserInfo[disperserId];
+
+        // Assert that the disperser address is non-zero.
+        if (disperser == address(0)) revert InputAddressZero();
+        // Assert that the disperser is not already registered.
+        if (disperserInfo.disperser != address(0)) revert AlreadyRegistered();
+
+        // Set the disperser info.
+        disperserInfo.disperser = disperser;
+        disperserInfo.relayURL = relayURL;
+
+        emit DisperserRegistered(disperserId, disperser);
+    }
+
+    /// @inheritdoc IEigenDADisperserRegistryV2
+    function deregisterDisperser(uint32 disperserId, bytes memory signature) external virtual {
+        EigenDATypesV2.DisperserInfoV2 storage disperserInfo = _disperserInfo[disperserId];
+
+        bytes32 digest = keccak256(abi.encode(DEREGISTRATION_TYPEHASH, disperserId, nonces[disperserInfo.disperser]++));
+
+        // Assert that the disperser is not in the default or on-demand dispersers sets.
+        // This means owner can only deregister dispersers after revoking their default or on-demand status.
+        if (_defaultDispersers.contains(disperserId) || _onDemandDispersers.contains(disperserId)) {
+            revert DisperserInSet();
+        }
+
+        _checkSignature(disperserInfo.disperser, digest, signature);
+        // Assert that the disperser is registered.
+        if (disperserInfo.disperser == address(0)) revert NotRegistered();
```

Comment:
Since it's impossible to create a valid signature from the zero address, I think this is technically unnecessary.